src/Microsoft.AspNetCore.Blazor.Browser.JS/src/Interop/InvokeDotNetMethodWithJsonMarshalling.ts

```diff
@@ -0,0 +1,183 @@
+  succeeded: boolean;
+  result?: any;
+  message?: string;
+}
+
+export interface MethodInstance {
+  name: string;
+  typeArguments?: { [key: string]: TypeInstance }
+  parameterTypes?: TypeInstance[];
+}
+
+export interface TypeInstance {
+  assembly: string;
+  name: string;
+  typeArguments?: { [key: string]: TypeInstance };
+}
+
+export function invokeDotNetMethod<T>(methodOptions: MethodOptions, ...args: any[]): (T | null) {
+  return invokeDotNetMethodCore(methodOptions, null, ...args);
+}
+
+let registrations = {};
+
+function resolveRegistration(methodOptions: MethodOptions) {
+  let existing = registrations[methodOptions.type.assembly];
+  existing = existing && registrations[methodOptions.type.name];
+  existing = existing && registrations[methodOptions.method.name];
+  if (existing !== undefined) {
+    return existing;
+  } else {
+    const method = platform.findMethod(
+      'Microsoft.AspNetCore.Blazor.Browser',
+      'Microsoft.AspNetCore.Blazor.Browser.Interop',
+      'JavaScriptInvoke',
+      'FindDotNetMethod');
+
+    const serializedOptions = platform.toDotNetString(JSON.stringify(methodOptions));
+    const result = platform.callMethod(method, null, [serializedOptions]);
+    const registration = platform.toJavaScriptString(result as System_String);
+
+    if (registrations[methodOptions.type.assembly] === undefined) {
+      let assembly = {};
+      let type = {};
+      registrations[methodOptions.type.assembly] = assembly;
+      assembly[methodOptions.type.name] = type;
+      type[methodOptions.method.name] = registration;
+    } else if (registrations[methodOptions.type.assembly][methodOptions.type.assembly] === undefined) {
+      let type = {};
+      registrations[methodOptions.type.assembly][methodOptions.type.name] = type;
+      type[methodOptions.type.name] = type;
+      type[methodOptions.method.name] = registration;
+    } else {
+      registrations[methodOptions.type.assembly][methodOptions.type.name][methodOptions.method.name] = registration;
+    }
+
+    return registration;
+  }
+}
+
+function invokeDotNetMethodCore<T>(methodOptions: MethodOptions, callbackId: string | null, ...args: any[]): (T | null) {
+  const method = platform.findMethod(
+    'Microsoft.AspNetCore.Blazor.Browser',
+    'Microsoft.AspNetCore.Blazor.Browser.Interop',
+    'JavaScriptInvoke',
+    'InvokeDotNetMethod');
+
+  const registration = resolveRegistration(methodOptions);
+
+  const packedArgs = packArguments(args);
+
+  const serializedCallback = callbackId != null ? platform.toDotNetString(callbackId) : null;
+  const serializedArgs = platform.toDotNetString(JSON.stringify(packedArgs));
+  const serializedRegistration = platform.toDotNetString(registration);
+  const serializedResult = platform.callMethod(method, null, [serializedRegistration, serializedCallback, serializedArgs]);
+
+  if (serializedResult !== null && serializedResult !== undefined && (serializedResult as any) !== 0) {
```

Comment:
It looks like the .NET-side `InvokeDotNetMethod` always returns a non-null string. As such, it doesn't look like the `if` test here is necessary, nor is the code path on line 97.

If I'm correct about this, can we remove the unused checks and code path? We don't want to silently swallow unexpected outcomes like the .NET side somehow returning `null` even though it's specifically coded not to.